### PR TITLE
[core] Include Abort[Nothing] in IO

### LINKED ISSUE #836: Consider subsuming `Abort[Nothing]` under `IO`
context:
 * https://github.com/getkyo/kyo/pull/815#issuecomment-2457651912

@fwbrasil 
> >    Abort[Nothing] corresponds to a fatal, unexpected failure/defect/bug, right?
>
>Yes, it represents a Result.Panic.
>
>>    In that case, shouldn't it already be subsumed by IO itself?
>
>That's a good point. It seems it'd be more convenient but then IO.Unsafe.run will need to return a Result. Do you mind creating an issue for it so we can follow up separately? It'll be trickier to implement since IO doesn't handle Abort currently.

### PR DESCRIPTION
**Pull request by fwbrasil**
Fixes #836

### CODE REVIEW

**Comment by fwbrasil** [comment]
@sideeffffect in case you have some time review

**Review by sideeffffect — APPROVED** [review]
Looks good from what I can tell :+1: Thanks

-----
A question/thought not totally related to this PR:

What is the distinction between IO and Async (regardless of naming) good for for a _user_ of Kyo?
I understand that this may be very important for the Kyo implementers. But do users really care about distinguishing between the two?

**Comment by fwbrasil** [comment]
> What is the distinction between IO and Async (regardless of naming) good for for a _user_ of Kyo? I understand that this may be very important for the Kyo implementers. But do users really care about distinguishing between the two?

We discussed this topic in https://github.com/getkyo/kyo/issues/709. The difference between a computation with a pending `IO` and one with `Async` is that the former indicates it can't park execution while the latter allows async boundaries. This distinction is quite useful for reasoning about the performance characteristics of computations. For instance, if you see a method in Kyo returning a pending `IO` without `Async`, you can safely assume the code will execute to completion and won't park nor lock, as Kyo's primitives are lock-free. This prevents surprises like Java's `URL.equals` making network calls for example.

That said, unless high performance is a concern, the distinction between `IO` and `Async` probably won't be relevant for most users. In practice, it's often possible to use `Async` directly since it includes `IO`.

File: kyo-core/shared/src/main/scala/kyo/IO.scala [hunk 1]
```diff
@@ -115,7 +133,7 @@ object IO:
           * @return
           *   The result of the IO effect, evaluated lazily along with its side effects.
           */
-        def runLazy[A: Flat, S](v: => A < (IO & S))(using Frame, AllowUnsafe): A < S =
+        def runLazy[E, A: Flat, S](v: => A < (IO & Abort[E] & S))(using Frame, AllowUnsafe): A < (S & Abort[E]) =
```
[the remark this answers is not in the document]
Comment (fwbrasil) [reply kyo-core/shared/src/main/scala/kyo/IO.scala:136]: Good point! Removing

**Comment by calvinlfer** [comment]
> For instance, if you see a method in Kyo returning a pending IO without Async, you can safely assume the code will execute to completion and won't park nor lock, as Kyo's primitives are lock-free.

this was super helpful and well written - it should be in the docs 🙏

**Pull request merged**